conversation_reconstruction/dataflow/conversation_reconstruction/shard.py

```diff
@@ -0,0 +1,148 @@
+import logging
+import subprocess
+import json
+from os import path
+import urllib2
+import traceback
+from google.cloud import bigquery as bigquery_op 
+from avro import schema, datafile, io
+import copy
+import sys
+
+import apache_beam as beam
+from apache_beam.options.pipeline_options import PipelineOptions
+from apache_beam.options.pipeline_options import SetupOptions
+from apache_beam.io.avroio import ReadFromAvro 
+from apache_beam.io import filesystems
+
+
+LOG_INTERVAL = 100
+
+SCHEMA_STR = """{
+    "type": "record",
+    "name": "ingested-sharded",
+    "namespace": "AVRO",
+    "fields": [
+        {   "name": "sha1",           "type": ["string" ,"null"]   },
+        {   "name": "user_id",        "type": ["string" ,"null"]   },
+        {   "name": "format",         "type": ["string" ,"null"]   },
+        {   "name": "user_text",      "type": ["string" ,"null"]   },
+        {   "name": "timestamp",      "type": ["string" ,"null"]   },
+        {   "name": "text",           "type": ["string" ,"null"]   },
+        {   "name": "page_title",     "type": ["string" ,"null"]   },
+        {   "name": "model",          "type": ["string" ,"null"]   },
+        {   "name": "page_namespace", "type": ["string" ,"null"]   },
+        {   "name": "page_id",        "type": ["string" ,"null"]   },
+        {   "name": "rev_id",         "type": ["string" ,"null"]   },
+        {   "name": "comment",        "type": ["string" ,"null"]   },
+        {   "name": "user_ip",        "type": ["string" ,"null"]   },
+        {   "name": "truncated",      "type": ["boolean","null"]   },
+        {   "name": "records_count",  "type": ["int"    ,"null"]   },
+        {   "name": "record_index",   "type": ["int"    ,"null"]   },
+        {   "name": "week",           "type": ["int"    ,"null"]   },
+        {   "name": "year",           "type": ["int"    ,"null"]   },
+        {   "name": "rev_id_in_int",  "type": ["int"    ,"null"]   }
+    ]
+}"""
+ 
+SCHEMA = schema.parse(SCHEMA_STR)
+THERESHOLD = 5000
+
+def run(known_args, pipeline_args):
+  """Main entry point; defines and runs the sharding pipeline."""
+
+  pipeline_args.extend([
+    '--runner=DataflowRunner',
+    '--project=wikidetox-viz',
+    '--staging_location=gs://wikidetox-viz-dataflow/staging',
+    '--temp_location=gs://wikidetox-viz-dataflow/tmp',
+    '--job_name=shard-{table}'.format(table=known_args.category),
+    '--num_workers=30'])
+  pipeline_options = PipelineOptions(pipeline_args)
+  pipeline_options.view_as(SetupOptions).save_main_session = True
+  jobname = 'shard-{table}'.format(table=known_args.category)
+
+  # Queries extracting the data
+  with beam.Pipeline(options=pipeline_options) as p:
+       pcoll = (p | ReadFromAvro(known_args.input) 
+                 | beam.Map(lambda x: ('{week}at{year}'.format(week=x['week'], year=x['year']), x)) 
+                 | beam.GroupByKey()
+                 | beam.ParDo(WriteToStorage()))
+
+class WriteToStorage(beam.DoFn):
```

Comment:
in general, writing to files is discouraged in DoFn's. It's better to use Filter/Partition and Beam IO libraries.
But there are exceptions. This is fine for now with caveats.